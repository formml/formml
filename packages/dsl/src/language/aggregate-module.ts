import { EmptyFileSystem, inject } from 'langium'
import {
  DefaultSharedModuleContext,
  LangiumSharedServices,
  createDefaultModule,
  createDefaultSharedModule,
} from 'langium/lsp'

import { FormMLModule, FormMLServices } from './formml/formml-module.js'
import { registerFormMLValidationChecks } from './formml/formml-validator.js'
import {
  FormMLDeclarationModule,
  FormMLDeclarationServices,
} from './formml-declaration/formml-declaration-module.js'
import {
  FormMLDeclarationGeneratedModule,
  FormMLGeneratedModule,
  FormMLGeneratedSharedModule,
} from './generated/module.js'
import LongestMatchServiceRegistry from './overrides/LongestMatchServiceRegistry.js'

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createAggregateServices(context: DefaultSharedModuleContext): {
  FormML: FormMLServices
  FormMLDeclaration: FormMLDeclarationServices
  shared: LangiumSharedServices
} {
  const shared = inject(
    createDefaultSharedModule(context),
    FormMLGeneratedSharedModule,
    {
      ServiceRegistry: () => new LongestMatchServiceRegistry(),
    },
  )
  const FormML = inject(
    createDefaultModule({ shared }),
    FormMLGeneratedModule,
    FormMLModule,
  )
  const FormMLDeclaration = inject(
    createDefaultModule({ shared }),
    FormMLDeclarationGeneratedModule,
    FormMLDeclarationModule,
  )
  shared.ServiceRegistry.register(FormML)
  shared.ServiceRegistry.register(FormMLDeclaration)

  registerFormMLValidationChecks(FormML)
  // registerFormMLDeclarationValidationChecks(FormMLDeclaration)

  if (!context.connection) {
    // We don't run inside a language server
    // Therefore, initialize the configuration provider instantly
    void shared.workspace.ConfigurationProvider.initialized({})
  }
  return { FormML, FormMLDeclaration, shared }
}

export const createInMemoryAggregateServices = () =>
  createAggregateServices(EmptyFileSystem)
