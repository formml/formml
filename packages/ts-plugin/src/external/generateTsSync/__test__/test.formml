form testForm {
  num  numField
  text textField
}
