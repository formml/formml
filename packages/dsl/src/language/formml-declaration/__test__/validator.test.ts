import { clearDocuments, parseHelper } from 'langium/test'

import type { FormMLDeclaration } from '../../generated/ast.js'

import { createInMemoryAggregateServices } from '../../aggregate-module.js'

describe('formml-declaration validator', () => {
  const services = createInMemoryAggregateServices()

  afterEach(async () => {
    await clearDocuments(services.shared)
  })

  const parser = (input: string) =>
    parseHelper<FormMLDeclaration>(services.FormMLDeclaration)(input, {
      validation: true,
    })

  describe('annotation', () => {
    test('invalidate required parameters after optional parameters', async () => {
      const input = 'annot fun foo(a: num, b?: text, c: bool)'
      const { diagnostics } = await parser(input)
      expect(diagnostics).toHaveLength(1)
      expect(diagnostics).toMatchSnapshot()
    })
  })
})
