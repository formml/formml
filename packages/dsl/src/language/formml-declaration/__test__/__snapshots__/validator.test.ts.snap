// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formml-declaration validator > function > invalidate required parameters after optional parameters 1`] = `
[
  {
    "code": undefined,
    "codeDescription": undefined,
    "data": undefined,
    "message": "A required parameter cannot follow an optional parameter",
    "range": {
      "end": {
        "character": 33,
        "line": 0,
      },
      "start": {
        "character": 26,
        "line": 0,
      },
    },
    "relatedInformation": undefined,
    "severity": 1,
    "source": "formml-declaration",
    "tags": undefined,
  },
]
`;
